File: GoTvApps/GoTvViewerAndroid/app/src/main/java/com/p2pgotv/gotvstation/ActivityCreateProfile.java
//============================================================================
// Copyright (C) 2011 Brett R. Jones
// All Rights Reserved ( After My Death I Brett R. Jones Issue this source code to Public Domain )
//
// You may not modify or redistribute this code for any reason
// without written consent of Brett R. Jones
//
// This code is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// brett1900@usa.com
// http://www.gotvp2p.net
//============================================================================

package com.p2pgotv.gotvstation;

import android.content.*;
import android.util.*;
import android.view.*;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import java.util.*;
import android.os.Bundle;

public class ActivityCreateProfile extends ActivityBase 
{
	public static final String LOG_TAG = "ActivityCreateProfile:";
	
	//=== vars ===//
	private UserAccount			m_oUserAccount; 
	private DataHelper			m_oDataHelper;
	// buttons
	private Button 				m_oLoginBut = null;
	private Button 				m_oCancelBut = null;
	// edit text
	private EditText			m_oUserName = null;
	private EditText			m_oMoodMsg = null; 	
	
    //========================================================================
    @Override
    public void onCreate(Bundle savedInstanceState) 
    {
        // launch Login activity
        Log.i(LOG_TAG, "onCreateCalled");
        
        setContentView(R.layout.activity_create_profile);
        super.onCreate(savedInstanceState);
        setPageTitle( R.string.activity_create_profile_name );
        
        m_oDataHelper = m_MyApp.getDataHelper();
        m_oUserAccount = m_MyApp.getUserAccount();
         
    	this.m_oLoginBut = (Button)this.findViewById(R.id.create_profile_login_button);
	 	this.m_oLoginBut.setOnClickListener(new OnClickListener() 
	 	{
	 		public void onClick(View v) 
	 		{
 			    m_MyApp.playButtonClick();
	 			OnLoginButClick(v);
	 		}
	 	});	 	
	 	
    	this.m_oCancelBut = (Button)this.findViewById(R.id.cancel_button);
	 	this.m_oCancelBut.setOnClickListener(new OnClickListener() 
	 	{
	 		public void onClick(View v) 
	 		{
 			    m_MyApp.playButtonClick();
	 			OnCancelButClick(v);
	 		}
	 	});	 	

	 	this.m_oMoodMsg = (EditText)this.findViewById(R.id.create_profile_mood_text);
	 	this.m_oUserName = (EditText)this.findViewById(R.id.create_profile_username_text);      
    }
    
    //========================================================================
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) 
    {
    	super.onActivityResult(requestCode, resultCode, data);
    	if( Constants.ACTIVITY_EDIT_PROFILE == requestCode )
    	{
    		// get intent for this activity
    		Intent oThisIntent = getIntent();
    		// return result of edit activity back to MainActivity
    		setResult(resultCode, oThisIntent);
    		m_MyApp.setUserAccount( this.m_oUserAccount );
    		finish();
    	}
    	
		setCamCaptureActivity( this );
    }
    
    //========================================================================
    //! Implement the OnClickListener callback    
	public void OnLoginButClick(View v) 
    {
    	if( true == AccountValidate() )
    	{
    		UUID uuidTmp = UUID.randomUUID(); 
       	 	m_oUserAccount.m_u64IdHiPart = uuidTmp.getMostSignificantBits();
       	 	m_oUserAccount.m_u64IdLoPart = uuidTmp.getLeastSignificantBits();
       	 	m_oUserAccount.m_as8OnlineName = m_oUserName.getText().toString();
       	 	m_oUserAccount.m_as8OnlineDesc = m_oMoodMsg.getText().toString();
       	 	m_oUserAccount.setDefaultPermissions();
       	 	int onlinePort = NativeTxTo.fromGuiGetRandomTcpPort();
       	 	m_oUserAccount.setOnlinePort( onlinePort );
       	 	m_MyApp.getMySettings().setMyTcpInPort( onlinePort );
       	 	Log.i(LOG_TAG, "created acct name " + m_oUserAccount.m_as8OnlineName + " UUID " + m_oUserAccount.m_u64IdHiPart + ", " + m_oUserAccount.m_u64IdLoPart);
       	 	if( true == m_MyApp.saveUserAccountToDatabase(m_oUserAccount))
       	 	{
           	 	if( true == m_oDataHelper.updateLastLogin(m_oUserAccount.m_as8OnlineName))
           	 	{
           	 		m_MyApp.setUserAccount( this.m_oUserAccount );
           	 		m_MyApp.myAppInitialize( this, getResources());
					String userXferDir 	= m_MyApp.getAppUsersDir() + m_oUserAccount.m_as8OnlineName + "/";
					m_MyApp.setUserXferDirectory( userXferDir );					

					String userSpecificDir 	= m_MyApp.getAppAccountsDir() + m_oUserAccount.m_as8OnlineName + "/";
					m_MyApp.setUserSpecificDirectory( userSpecificDir );
					NativeTxTo.fromGuiUpdateMyIdent( m_oUserAccount, false );
           	 		
           	 		// give user a chance to edit his/her profile
           	    	Intent myIntent = new Intent(ActivityCreateProfile.this, ActivityEditProfile.class);
           	    	ActivityCreateProfile.this.startActivityForResult(myIntent, Constants.ACTIVITY_EDIT_PROFILE );
           	 	}		
           	 	else
           	 	{
           	 		Toast.makeText(this, "Update New User into Database failed.", Toast.LENGTH_LONG).show();
           	 		Log.i(LOG_TAG, "Update New User into Database failed.");
           	 	}
       	 	}
       	 	else
       	 	{
       	 		Toast.makeText(this, "Insert New User into Database failed.", Toast.LENGTH_LONG).show();
       	 		Log.i(LOG_TAG, "Insert New User into Database failed.");
       	 	}
    	}   	
    }   
	
    //========================================================================
    public boolean AccountValidate()  
    {
	   	 String strUserName = m_oUserName.getText().toString();
	   	 if( strUserName.contains("'") )
	   	 {
			 Toast.makeText(this, "User Name cannot have special character quote.", Toast.LENGTH_LONG).show();;
			 return false;   		 
	   	 }
	   
	   	 /*
	 	if( strUserName.contains("My P2P Web") 
	 			||  strUserName.contains("MyP2PWeb")
	 			||  strUserName.contains("myp2pweb")
	 			||  strUserName.contains("my p2p web") )
	   	 {
			 Toast.makeText(this, "User Name cannot contain name My P2P Web.", Toast.LENGTH_LONG).show();;
			 return false;   		 
	   	 }*/
   	 
	   	 if( strUserName.contains("(") || strUserName.contains(")") )
	   	 {
			 Toast.makeText(this, "User Name cannot have special character parentheses.", Toast.LENGTH_LONG).show();;
			 return false;   		 
	   	 }
	   	 
	   	 if( strUserName.contains("/") || strUserName.contains("\\") )
	   	 {
			 Toast.makeText(this, "User Name cannot have special character slashes.", Toast.LENGTH_LONG).show();;
			 return false;   		 
	   	 }
	   	 
		 if( strUserName.length() > 23 )
		 {
			 Toast.makeText(this, "User Name is too long (max 23 chars).", Toast.LENGTH_LONG).show();;
			 return false;
		 }
		 
		 if( strUserName.length() < 4 )
		 {
			 Toast.makeText(this, "User Name is too short (min 4 chars).", Toast.LENGTH_LONG).show();;
			 return false;
		 }
		 
    	 String strMoodMsg = m_oMoodMsg.getText().toString();
	   	 if( strUserName.contains("'") )
	   	 {
			 Toast.makeText(this, "Mood Message cannot have special character quote.", Toast.LENGTH_LONG).show();;
			 return false;   		 
	   	 }
	   	 
	   	 int iLen = strMoodMsg.length();
	   	 if( iLen > 27 )
    	 {
    		 Toast.makeText(this, "Mood Message is too long (max 27 chars)", Toast.LENGTH_LONG).show();;
    		 return false;
    	 }
	   	 
    	 return true;
    }
    
    //========================================================================
 	public void OnCancelButClick(View v) 
    {
		onBackButtonPressed();
	}
    
 	//========================================================================
    @Override
    public void onBackButtonPressed() 
    {
		Intent intent = this.getIntent(); 
		//intent.putExtra("SOMETHING", "EXTRAS"); 
		this.setResult(RESULT_CANCELED, intent); 
		super.onBackButtonPressed();
	}
}

File: GoTvApps/GoTvCommon/QtSource/ActivityCreateProfile.cpp
//============================================================================
// Copyright (C) 2010 Brett R. Jones
// Issued to MIT style license by Brett R. Jones in 2017
//
// You may use, copy, modify, merge, publish, distribute, sub-license, and/or sell this software
// provided this Copyright is not modified or removed and is included all copies or substantial portions of the Software
//
// This code is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// bjones.engineer@gmail.com
// http://www.gotvptop.net
//============================================================================

#include "ActivityCreateProfile.h"
#include "ActivityEditProfile.h"
#include "AppGlobals.h"
#include "AppCommon.h"
#include "VxDataHelper.h"
#include "AppSettings.h"

#include <GoTvCore/GoTvP2P/P2PEngine/EngineSettings.h>
#include <CoreLib/VxFileUtil.h>
#include <CoreLib/VxGlobals.h>

#include <QMessageBox>
#include <QUuid>

//============================================================================
ActivityCreateProfile::ActivityCreateProfile( AppCommon& app, QWidget * parent )
: ActivityBase( OBJNAME_ACTIVITY_CREATE_PROFILE, app, parent, eActivityCreateProfile )
{
	ui.setupUi( getContentItemsFrame() );
	connect( ui.loginButton, SIGNAL(clicked()), this, SLOT(buttonLoginClicked()));
}

//============================================================================
ActivityCreateProfile::~ActivityCreateProfile()
{
}

//============================================================================
//! button clicked
void ActivityCreateProfile::buttonLoginClicked( void )
{
	if( true == accountValidate() )
	{
		std::string strUserName = ui.userNameEdit->text().toUtf8().data();

		m_MyApp.loadAccountSpecificSettings( strUserName.c_str() );
		m_MyApp.createAccountForUser( strUserName, m_UserAccount, (const char *) ui.moodMessageEdit->text().toUtf8().data() );
		m_MyApp.setAccountUserName( strUserName.c_str() );

		// save ident to global
		memcpy( m_MyApp.getAppGlobals().getUserIdent(), &m_UserAccount, sizeof( VxNetIdent ));

		// fill in database info
		LogMsg( LOG_VERBOSE, "created acct name %s UUID 0x%llX 0x%llX\n", m_UserAccount.getOnlineName(), m_UserAccount.getMyOnlineIdLoPart(), m_UserAccount.getMyOnlineIdHiPart());
		if( true == m_MyApp.getDataHelper().insertAccount( m_UserAccount ) )
		{
			m_MyApp.getDataHelper().updateLastLogin( m_UserAccount.getOnlineName() );

			// give user a chance to edit his/her profile
			ActivityEditProfile oDlg( m_MyApp, VxGetUserSpecificDataDirectory().c_str(), 0 );
			if( QDialog::Rejected == oDlg.exec() )
			{
				reject();
			}
			else
			{
				accept();
			}
		}
		else
		{
			LogMsg( LOG_ERROR, "Insert New User into Database failed.");
			QMessageBox::warning(this, tr("Application"), tr("Insert New User into Database failed." ) );
			reject();
		}
	}   	
}

//============================================================================
//! validate user input
bool ActivityCreateProfile::accountValidate( void )
{
	QString strUserName = ui.userNameEdit->text();
	if( strUserName.contains("GoTv PtoP Web") 
		|| strUserName.contains("GoTvPtoPWeb")
		|| strUserName.contains("gotvptopweb")
		|| strUserName.contains("gotv ptop web")
        || strUserName.contains( "GoTv PtoP Web" )
        || strUserName.contains( "GoTvPtoPWeb" )
        || strUserName.contains( "gotvptopweb" )
        || strUserName.contains( "gotv ptop web" ) )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr("User Name cannot have GoTv PtoP Web in name." ) );
		return false;   		 
	}

	if( strUserName.contains("'") ||  strUserName.contains("\"") )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr("User Name cannot have special character quote." ) );
		return false;   		 
	}

	if( strUserName.contains(",") )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr("User Name cannot have comma." ) );
		return false;   		 
	}

	if( strUserName.contains("(") || strUserName.contains(")") )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr( "User Name cannot have special character parentheses." ) );
		return false;   		 
	}

	if( strUserName.contains("/") || strUserName.contains("\\") )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr("User Name cannot have special character slashes." ) );
		return false;   		 
	}

	if( strUserName.length() > 23 )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr("User Name is too long (maximum 23 chars)." ) );
		return false;
	}

	if( strUserName.length() < 4 )
	{
		QMessageBox::warning(this, QObject::tr("Invalid User Name"), QObject::tr("User Name is too short (minimum 4 chars)." ) );
		return false;
	}

	QString strMoodMsg = ui.moodMessageEdit->text();
	if( strUserName.contains("'") )
	{
		QMessageBox::warning(this, QObject::tr("Application"), QObject::tr("Mood Message cannot have special character quote." ) );
		return false;   		 
	}

	int iLen = strMoodMsg.length();
	if( iLen > 27 )
	{
		QMessageBox::warning(this, QObject::tr("Application"), QObject::tr("Mood Message is too long (maximum 27 chars)" ) );
		return false;
	}

	return true;
}

//============================================================================
bool ActivityCreateProfile::wasLoginNameEntered( void )
{
    QString strUserName = ui.userNameEdit->text();
    return strUserName.length() > 0 ? true : false;
}

